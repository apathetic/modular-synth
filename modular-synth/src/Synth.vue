//------------------------------------------------
//  APPLICATION
// -----------------------------------------------

<template>
  <section :class="editing ? 'edit-mode': 'play-mode'">
    <div
      id="modules"
      class="grid-container"
      ref="grid"
      @click.left="clearActive"
      >

      <div class="position-highlight">
        <div class="inner"></div>
      </div>

      <component v-for="(module, index) in modules"
        :is="module.type"
        :id="module.id"
        :col="module.col"
        :row="module.row"
        :coords="{ x:module.x, y:module.y }"
        :key="module.id"

        @mousedown.native="setActive(module.id)"
        @mouseover.native="setFocus(module.id)"
        @mouseout.native="clearFocus">
      </component>

      <svg id="connections">
        <connecting></connecting>
        <connection v-for="connection in connections"
          :id="connection.id"
          :to="connection.to"
          :from="connection.from"
          :key="connection.id"

          @mousedown.native="setActive(connection.id)">
        </connection>
      </svg>
    </div>

    <aside id="sidebar">
      <div class="controls pad">
        <h4>{{ editing ? 'EDIT MODE' : 'PERFORMANCE MODE' }}</h4>

        <button class="mode" @click="toggleEditMode">
          <span class="play">play</span>
          <span class="edit">edit</span>
        </button>

        <p v-if="active">
          <strong>Current Module</strong><br>
          {{ active.type }} (id: {{ active.id }})<br>
          x, y: {{ active.x }}, {{ active.y }}<br>
          col, row: {{ active.col }}, {{ active.row }}<br>
          w, h: {{ active.w }},  {{ active.h }}<br>
        </p>

        <midi></midi>

        <br>

        <!-- <button
          class="power"
          :class="power ? 'on' : 'off'"
          @click="togglePower">

          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 40">
            <path d="M28,18c0,6.629-5.375,12-12,12C9.371,30,4,24.629,4,18c0-5.223,3.34-9.652,8-11.301v4.41C9.617,12.496,8,15.047,8,18 c0,4.418,3.582,8,8,8s8-3.582,8-8c0-2.953-1.621-5.504-4-6.891v-4.41C24.656,8.348,28,12.777,28,18z M16,16c1.105,0,2-0.895,2-2V4 c0-1.104-0.895-2-2-2s-2,0.896-2,2v10C14,15.105,14.895,16,16,16z" />
          </svg>
        </button> -->
      </div>

      <master-out></master-out>

      <div class="power pad">
        <button
          :class="power ? 'on' : 'off'"
          @click="togglePower">

          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32">
            <path d="M28,18c0,6.629-5.375,12-12,12C9.371,30,4,24.629,4,18c0-5.223,3.34-9.652,8-11.301v4.41C9.617,12.496,8,15.047,8,18 c0,4.418,3.582,8,8,8s8-3.582,8-8c0-2.953-1.621-5.504-4-6.891v-4.41C24.656,8.348,28,12.777,28,18z M16,16c1.105,0,2-0.895,2-2V4 c0-1.104-0.895-2-2-2s-2,0.896-2,2v10C14,15.105,14.895,16,16,16z" />
          </svg>
        </button>
      </div>
    </aside>

  </section>
</template>

<script>
  import { mapGetters, mapActions } from 'vuex';
  import { sortable } from './mixins/sortable';

  import Analyser from './components/Analyser';
  import Comb from './components/Comb';
  import Delay from './components/Delay';
  import Env from './components/Env';
  import LFO from './components/LFO';
  import Mixer from './components/Mixer';
  import NoteIn from './components/NoteIn';
  import Reverb from './components/Reverb';
  import VCF from './components/Filter';
  import VCO from './components/VCO';
  import VCA from './components/VCA';

  import Debugger from './components/Debugger';
  import Node from './components/Node';

  import connecting from './components/system/Connecting';
  import connection from './components/system/Connection';
  import masterOut from './components/system/MasterOut';
  import midi from './components/system/Midi.vue';


  export default {
    mixins: [sortable],

    components: {
      masterOut,
      connecting,
      connection,
      midi,

      Env,
      LFO,
      VCF,
      VCO,
      VCA,
      Analyser,
      Comb,
      Debugger,
      Delay,
      Mixer,
      Node,
      NoteIn,
      Reverb
    },

    computed: {
      ...mapGetters([
        'power',
        'editing',
        'active',
        'modules',
        'connections'
      ])
    },

    data() {
      return {
        sorting: false
      };
    },

    created() {
      console.log('◌ App: loading...');

      this.$bus.$on('drag:start', (coords, el) => {
        if (!this.editing) {
          this.startSorting();   // from sortable mixin
        }
      });

      this.$bus.$on('drag:active', (coords, el) => {
        if (!this.editing) { //  this.sorting) {
          this.whileSorting(el); // from sortable mixin
        }
      });

      this.$bus.$on('drag:end', () => {
        if (!this.editing) {
          this.stopSorting();    // from sortable mixin
        }
      });

      this.$bus.$on('module:add', () => {
        console.log('module add');
        this.$nextTick(function() {
          const item = this.modules.slice(-1)[0]; // get last (newest) item

          this.gridList.items = this.modules;
          this.gridList.moveItemToPosition(item, [0, 0]);
        });
      });

      this.$bus.$on('module:remove', () => {
        console.log('module remove');
        this.$nextTick(() => {
          // const deleted = this.modules[this.$store.state.active];
          // this.gridList._deleteItemPositionFromGrid(deleted);

          this.gridList.items = this.modules;
          this.gridList._pullItemsToLeft();
        });
      });

      window.addEventListener('keydown', (e) => {
        switch (e.code) {
          case 'Delete':
          case 'Backspace':
            this.removeModule();
            this.$bus.$emit('module:remove');
            break;
          case 'Tab':
            this.toggleEditMode();
            break;
          case 'Escape':
            // this.togglePower();
            break;
          case 'Space':
            // this.togglePlay();  // one day, if i plug in a timeline
            break;
          case 'ShiftLeft':
          case 'ShiftRight':
            // WE only want to rearrange the module-rack if shift is held;
            // otherwise, we probably want to play the module
            // this.toggleSorting;
            this.sorting = true;
            break;
          default:
            // console.log(e.code);
        }
      });

      window.addEventListener('keyup', (e) => {
        switch (e.code) {
          case 'ShiftLeft':
          case 'ShiftRight':
            this.sorting = false;
            break;
        }
      });
    },

    mounted() {
      console.log('◌ App: mounting...');
      // TODO why cannot move into sortable:ready() ...? A: $refs.grid is not yet in the DOM
      // TODO2 sortable:mounted() ...?
      this.initSorting(this.$refs.grid);
      // this.setupGrid();  // this should be after every module has registered its dimensions
    },

    methods: {
      ...mapActions([
        'togglePower',
        'toggleEditMode',
        'removeModule',
        'setActive',
        'clearActive',
        'setFocus',
        'clearFocus'
      ])
    }
  };
</script>

<style lang="scss">
  @import 'assets/scss/variables.scss';
  @import 'assets/scss/styles.scss';
  @import 'assets/scss/module.scss';
</style>
