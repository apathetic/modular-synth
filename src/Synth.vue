<template>
  <section :class="editing ? 'edit-mode': 'play-mode'">

    <div id="modules" ref="grid" @click.left="clearActive">
      <div class="position-highlight">
        <div class="inner"></div>
      </div>

      <module v-for="(module, index) in modules"
        :module="module"
        :key="module.id"

        @mousedown.native="setActive(module.id)"
        @mouseover.native="setFocus(module.id)"
        @mouseout.native="clearFocus()">
      </module>

      <svg id="connections" :style="width">
        <connecting></connecting>
        <connection v-for="connection in connections"
          :id="connection.id"
          :to="connection.to"
          :from="connection.from"
          :key="connection.id"

          @mousedown.native="setActive(connection.id)">
        </connection>
      </svg>
    </div>

    <aside id="sidebar">
      <div class="controls pad">
        <h4>{{ editing ? 'EDIT MODE' : 'PERFORMANCE MODE' }}</h4>

        <button class="mode" @click="toggleEditMode">
          <span class="play">play</span>
          <span class="edit">edit</span>
        </button>

        <p v-if="active">
          <strong>Current Module</strong><br>
          {{ active.type }} (id: {{ active.id }})<br>
          x, y: {{ active.x }}, {{ active.y }}<br>
          col, row: {{ active.col }}, {{ active.row }}<br>
          w, h: {{ active.w }},  {{ active.h }}<br>
        </p>

        <midi></midi>

      </div>

      <master-out></master-out>

      <div class="power pad">
        <button
          :class="power ? 'on' : 'off'"
          @click="togglePower">

          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 32 32">
            <path d="M28,18c0,6.629-5.375,12-12,12C9.371,30,4,24.629,4,18c0-5.223,3.34-9.652,8-11.301v4.41C9.617,12.496,8,15.047,8,18 c0,4.418,3.582,8,8,8s8-3.582,8-8c0-2.953-1.621-5.504-4-6.891v-4.41C24.656,8.348,28,12.777,28,18z M16,16c1.105,0,2-0.895,2-2V4 c0-1.104-0.895-2-2-2s-2,0.896-2,2v10C14,15.105,14.895,16,16,16z" />
          </svg>
        </button>
      </div>
    </aside>

  </section>
</template>

<script>
  import { mapGetters, mapActions } from 'vuex';
  import { sortable } from './mixins/sortable';
  import { EVENT } from './events';

  import connecting from './components/system/Connecting';
  import connection from './components/system/Connection';
  import masterOut from './components/system/MasterOut';
  import midi from './components/system/Midi.vue';
  import module from './components/system/Module';

  export default {
    mixins: [sortable],

    components: {
      masterOut,
      connecting,
      connection,
      midi,
      module
    },

    computed: {
      width() {
        const canvasWidth = this.bounds + 124 + 40; // .. + module width + 40
        return this.editing
          ? `width: ${canvasWidth}px`
          : 'width: auto';
      },

      ...mapGetters([
        'power',
        'editing',
        'active',
        'modules',
        'bounds',
        'connections'
      ])
    },

    data() {
      return {
        sorting: false
      };
    },

    created() {
      console.log('%c â—Œ Synth: loading... ', 'background:black;color:white;font-weight:bold');

      this.$bus.$on(EVENT.DRAG_START, (coords, el) => {
        if (!this.editing) {
          this.startSorting(); // from sortable mixin
        }
      });

      this.$bus.$on(EVENT.DRAG_ACTIVE, (coords, el) => {
        if (!this.editing) {
          this.whileSorting(el); // from sortable mixin
        }
      });

      this.$bus.$on(EVENT.DRAG_END, () => {
        if (!this.editing) {
          this.stopSorting(); // from sortable mixin
        }
      });

      this.$bus.$on(EVENT.APP_SORT, () => {
        this.initSorting(this.$refs.grid);
      });

      this.$bus.$on(EVENT.MODULE_ADD, () => {
        this.$nextTick(function() {
          const item = this.modules.slice(-1)[0]; // get last (newest) item

          this.gridList.items = this.modules;
          this.gridList.moveItemToPosition(item, [0, 0]);
        });
      });

      this.$bus.$on(EVENT.MODULE_REMOVE, () => {
        this.$nextTick(() => {
          this.gridList.items = this.modules;
          this.gridList._pullItemsToLeft();
        });
      });

      window.addEventListener(EVENT.KEY_DOWN, (e) => {
        switch (e.code) {
          case 'Delete':
          case 'Backspace':
            this.removeModule();
            this.$bus.$emit(EVENT.MODULE_REMOVE);
            break;
          case 'Tab':
            this.toggleEditMode();
            e.preventDefault(); // do not tab through <select>, fields, etc
            break;
          case 'Escape':
            // this.togglePower();
            break;
          case 'Space':
            // this.togglePlay();  // one day, if i plug in a timeline
            break;
          case 'ShiftLeft':
          case 'ShiftRight':
            // WE only want to rearrange the module-rack if shift is held;
            // otherwise, we probably want to play the module
            // this.toggleSorting;
            this.sorting = true;
            break;
          default:
            // console.log(e.code);
        }
      });

      window.addEventListener(EVENT.KEY_UP, (e) => {
        switch (e.code) {
          case 'ShiftLeft':
          case 'ShiftRight':
            this.sorting = false;
            break;
        }
      });
    },

    mounted() {
      const grid = this.$refs.grid; // rare time we need to scrape DOM.
      grid.addEventListener(EVENT.SCROLL, (e) => {
        if (this.editing) {
          this.$store.commit('UPDATE_SCROLL_OFFSET', e.target.scrollLeft);
        }
      });
    },

    methods: {
      ...mapActions([
        'togglePower',
        'toggleEditMode',
        'removeModule',
        'setActive',
        'clearActive',
        'setFocus',
        'clearFocus'
      ])
    }
  };
</script>

<style lang="scss">
  @import 'assets/scss/variables.scss';
  @import 'assets/scss/styles.scss';
  @import 'assets/scss/module.scss';

  #modules {
  // #canvas {
    flex: 1;
    overflow-x: auto;
    overflow-y: hidden;



    display:inline-block;
    width: auto;

    &::-webkit-scrollbar {
      width: 1em;
      height: 1em;
    }

    &::-webkit-scrollbar-track {
      background: $color-grey-dark;
    }

    &::-webkit-scrollbar-thumb {
      background-color: $color-grey-medium;
      border: 2px solid $color-grey-dark;
      border-radius: 0.5em;
    }

    &::-webkit-scrollbar-corner {
      background: $color-grey-dark;
    }
  }

  #modules {

  }

  #connections {
    min-width: 100%;
    min-height: 100%;
    position: absolute;
    top: 0;
    left: 0;
    transition: opacity 0.1s;
    // transition-delay: 0.2s;
    transition-delay: $transition-time-slow;

    .play-mode & {
      opacity: 0;
      z-index: -1;
      transition-delay: 0s;
    }
  }

  #sidebar {
    display: flex;
    flex-direction: column;
    flex-basis: 112px;
    background-color: #444;
    z-index: 9999;          // as nodes will increment their z-index
  }

  .controls {
    flex: 1;
  }
</style>
