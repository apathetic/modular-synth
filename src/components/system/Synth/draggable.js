/**
 * Draggable.
 * This mixin enables dragging of a Component. A position Object is updated in the Component's
 * data, and any/all connections to any inputs/outputs are also updated and redrawn.
 * @type {Object}
 */

import { reactive, computed, onMounted, ref, watch } from "@vue/composition-api";
import { useSortable } from './sortable';
import { rackWidth, rackHeight } from '../constants';
import { EVENT } from '../events';
// import  { useStore, useComputed } from 'vuex';

const dragObj = {
  zIndex: 0,
  cursorStartX: null,
  cursorStartY: null,
  startX: null,
  startY: null
};

export default {
  props: {
    module: Object
  },

  setup(props, { root }) {
    const { startSorting, whileSorting, stopSorting } = useSortable();
    const store = root.$store;
    const editing = computed(() => store.state.editing); // isEditing TODO
    const isDragging = false;
    const module = props.module;
    let element;

    const state = reactive({
      x: module.x,
      y: module.y,
      position: computed(() => ({
        left: `${ (editing || isDragging) ? state.x : module.col * rackWidth }px`,
        top:  `${ (editing || isDragging) ? state.y : module.row * rackHeight }px`
      })),
    });

    onMounted(() => {
      // the DOM element will be assigned to the ref after initial render
      console.log(root.value); // vm.$el
      element = root.value;
    });

    function startDragging(event) {
      const x = element.offsetLeft;  // Calculate explicity because could be in play mode, in which ...
      const y = element.offsetTop;   // case x,y would not pertain to the actual element coords.

      element.style.zIndex = ++dragObj.zIndex;     // Update element's z-index.

      // Save starting positions of cursor and element.
      dragObj.cursorStartX = event.clientX;
      dragObj.cursorStartY = event.clientY;
      dragObj.startX = x;
      dragObj.startY = y;

      isDragging = true;

      state.x = x;                 // necessary as the module's internal coords may be different if in play mode
      state.y = y;

      // this.$bus.$emit(EVENT.DRAG_START, [x, y], element);
      if (!editing) {
        startSorting();
      }

      // Capture mousemove and mouseup events on the page.
      document.addEventListener(EVENT.MOUSE_MOVE, whileDragging);
      document.addEventListener(EVENT.MOUSE_UP, stopDragging);
    }

    function whileDragging(event) {
      const x = dragObj.startX + event.clientX - dragObj.cursorStartX;
      const y = dragObj.startY + event.clientY - dragObj.cursorStartY;

      // we _always_ want to change the module's coordinates
      // while dragging, whether in editing mode or not:
      state.x = x;
      state.y = y;

      // this.$bus.$emit(EVENT.DRAG_ACTIVE, [x, y], element);
      whileSorting();
    }

    function stopDragging(event) {
      isDragging = false;
      // this.$bus.$emit(EVENT.DRAG_END, this.id);
      stopSorting();

      // we only want to update the Store with the
      // new coordinates if we are in editing mode:
      if (editing) {
        store.commit('UPDATE_GRID_POSITION', {
          id: module.id,
          x: state.x,
          y: state.y
        });
      } else {
        // otherwise, restore the x,y coordinates -- we don't want the
        // module to have moved around when we switch out of play mode
        const previous = store.getters['activeModule'];

        state.x = previous.x;
        state.y = previous.y;
      }

      document.removeEventListener(EVENT.MOUSE_MOVE, whileDragging);
      document.removeEventListener(EVENT.MOUSE_UP, stopDragging);
    }

    return {
      state,
      startDragging,
      whileDragging,
      stopDragging
    }
  }
};
