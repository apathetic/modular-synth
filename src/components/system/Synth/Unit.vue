// TODO: could this be a renderless component?
<template>
  <div
    class="module"
    :class="[width, tall, isDragging ? 'dragging' : '']"
    :style="position"
    @mousedown.stop="startDragging"
  >
    <slot></slot>
  </div>
</template>

<script>
  import { draggable } from '@/mixins/draggable';
  import { context } from '@/audio';
  import { moduleSize } from '@/constants';

  export default {
    mixins: [ draggable ],
    props: {
      module: Object
    },

    computed: {
      width() {
        return `_${moduleSize[this.module.type][0]}U`;
      },
      tall() {
        return moduleSize[this.module.type][2] ? 'module--tall' : '';
      }
    },

    created() {
      this.id = this.module.id; // NOTE: this is the ID used by the Connector to route audio

      // this.$store.commit('ADD_TO_REGISTRY', {
      //   id: this.id,
      //   node: this.$children[0], // this.$slots.default
      // });

      console.log('%c[component] Creating %s', 'color: green', this.module.type);
    },

    mounted() {
      this.$store.commit('app/ADD_TO_REGISTRY', {
        id: this.id,
        node: this.$children[0], // this.$slots.default
      });
    },

    destroyed() {
      console.log('Destroying %s ', this.module.type);
      this.$store.commit('REMOVE_FROM_REGISTRY', this.id);
    }
  };
</script>
