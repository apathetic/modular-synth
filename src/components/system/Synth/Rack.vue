<template>
  <div
    id="modules"
    ref="grid"
    XXXXv-contextmenu:modules
  >
    <div class="position-highlight">
      <div class="inner"></div>
    </div>

    <Unit
      v-for="module in modules"
      :module="module"
      :key="module.id"
      @XXXmousedown.stop="startDragging()"
      @mousedown.native="setActive(module.id)"
      @mouseover.native="setFocus(module.id)"
      @mouseout.native="clearFocus()"
    >
      <component
        :is="module.type"
        :id="module.id"
      />
    </Unit>

    <svg id="connections" :style="width">
      <connecting></connecting>
      <connection v-for="connection in connections"
        :id="connection.id"
        :to="connection.to"
        :from="connection.from"
        :key="connection.id"
        @mousedown.native="setActive(connection.id)">
      </connection>
    </svg>

  </div>
</template>

<script>
  import { context } from '@/audio';
  import { mapGetters, mapActions } from 'vuex';
  import { sortable } from '@/mixins/sortable';
  import { draggable } from '@/mixins/draggable';
  import { EVENT } from '@/events';

  import Connecting from '@/components/system/Connecting.vue';
  import Connection from '@/components/system/Connection.vue';
  import * as Modules from '@/components/';
  import Unit from './Unit';
  import Debugger from '@/components/test/Debugger';

  export default {
    name: 'Rack',
    provide: [ context ],
    mixins: [ sortable ], // draggable

    components: {
      ...Modules,
      Unit,
      Connecting,
      Connection,
      Debugger,
    },

    props: {
      modules: Array,
      connections: Array,
    },

    computed: {
      width() {
        const canvasWidth = this.bounds + 124 + 40; // .. + module width + 40
        return this.editing
          ? `width: ${canvasWidth}px`
          : 'width: auto';
      },
      ...mapGetters([
        'editing',
        'bounds',
      ])
    },

    created() {
      const bus = this.$root.$bus;
      bus.$on(EVENT.DRAG_START, (coords, el) => {
        if (!this.editing) {
          this.startSorting(); // from sortable mixin
        }
      });

      bus.$on(EVENT.DRAG_ACTIVE, (coords, el) => {
        if (!this.editing) {
          this.whileSorting(el); // from sortable mixin
        }
      });

      bus.$on(EVENT.DRAG_END, () => {
        if (!this.editing) {
          this.stopSorting(); // from sortable mixin
        }
      });

      bus.$on(EVENT.APP_SORT, () => {
        this.initSorting(this.$refs.grid);
      });

      bus.$on(EVENT.MODULE_ADD, () => {
        this.$nextTick(function() {
          const item = this.modules.slice(-1)[0]; // get last (newest) item

          this.gridList.items = this.modules;
          this.gridList.moveItemToPosition(item, [0, 0]);
        });
      });

      bus.$on(EVENT.MODULE_REMOVE, () => {
        this.$nextTick(() => {
          this.gridList.items = this.modules;
          this.gridList._pullItemsToLeft();
        });
      });
    },

    mounted() {
      // const grid = this.$refs.grid; // rare time we need to scrape DOM.
      // grid.addEventListener(EVENT.SCROLL, (e) => {
      //   if (this.editing) {
      //     this.$store.commit('UPDATE_SCROLL_OFFSET', e.target.scrollLeft);
      //   }
      // });
    },

    methods: {
      ...mapActions([
        'setActive',
        'clearActive',
        'setFocus',
        'clearFocus'
      ])
    }
  };
</script>
