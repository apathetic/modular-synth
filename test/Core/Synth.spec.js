import { shallow } from 'vue-test-utils'
import { createRenderer } from 'vue-server-renderer'
import Synth from '@/Synth.vue';


describe('Synth.vue', () => {

  it('matches snapshot', () => {
    const renderer = createRenderer()
    const app = shallow(Synth, {
      propsData: {}
    });

    renderer.renderToString(app.vm, (err, str) => {
      if (err) throw new Error(err);
      expect(str).toMatchSnapshot();
    });
  });


  it('should render correct contents', () => {
    const app = shallow(Synth, {
      propsData: {}
    });

    // expect(app.find('#modules').to.exist());
    // expect(app.find('#connections').to.exist());
    // expect(app.find('#aside').to.exist());
    // expect(app.find('#master-out').to.exist());
  });


  it('user can drag a module in edit mode', () => {});

  it('audio is routed correctly after load', () => {});

});
