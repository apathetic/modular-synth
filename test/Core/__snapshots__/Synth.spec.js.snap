// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Synth.vue matches snapshot 1`] = `<!---->`;
